src/features/tokens/types.ts

```diff
@@ -1,22 +1,83 @@
+import { ethers } from 'ethers';
 import { z } from 'zod';
 
-/**
- * Zod schema for Token config validation
- */
-export const TokenSchema = z.object({
-  chainId: z.number(),
-  address: z.string(),
-  hypCollateralAddress: z.string(),
+import { TokenType } from '@hyperlane-xyz/hyperlane-token';
+
+const commonTokenFields = z.object({
+  chainId: z.number().positive(),
   name: z.string(),
   symbol: z.string(),
   decimals: z.number().positive(),
   logoURI: z.string().optional(),
 });
+type CommonTokenFields = z.infer<typeof commonTokenFields>;
+
+/**
+ * Types for the developer-provided config
+ * See src/consts/tokens.ts
+ */
+
+interface BaseTokenConfig extends CommonTokenFields {
+  type: `${TokenType}`; // use template literal to allow string values
+}
+
+const CollateralTokenSchema = commonTokenFields.extend({
+  type: z.literal(TokenType.collateral),
+  address: z.string(),
+  hypCollateralAddress: z.string(),
+});
 
-export const TokenListSchema = z.array(TokenSchema);
+interface CollateralTokenConfig extends BaseTokenConfig {
+  type: TokenType.collateral | 'collateral';
+  address: Address;
+  hypCollateralAddress: Address;
+}
 
-export type TokenMetadata = z.infer<typeof TokenSchema>;
+const NativeTokenSchema = commonTokenFields.extend({
+  type: z.literal(TokenType.native),
+  hypNativeAddress: z.string(),
+});
 
-export interface TokenMetadataWithHypTokens extends TokenMetadata {
-  hypTokens: Array<{ chainId: number; address: Address }>;
+interface NativeTokenConfig extends BaseTokenConfig {
+  type: TokenType.native | 'native';
+  hypNativeAddress: Address;
 }
+
+export const WarpTokenConfigSchema = z.array(CollateralTokenSchema.or(NativeTokenSchema));
+export type WarpTokenConfig = Array<CollateralTokenConfig | NativeTokenConfig>;
+
+/**
+ * Types for use in the app after processing config
+ * Necessary to allow more flexibility with the config
+ * See src/features/tokens/metadata.ts
+ */
+interface BaseTokenMetadata extends CommonTokenFields {
+  type: TokenType;
+  address: Address;
+  hypWrapperAddress: Address; // Shared name for hypCollateralAddr and hypNativeAddr
+}
+
+interface CollateralTokenMetadata extends BaseTokenMetadata {
+  type: TokenType.collateral;
+}
+
+type ZeroAddress = `${typeof ethers.constants.AddressZero}`;
+interface NativeTokenMetadata extends BaseTokenMetadata {
```

Comment:
This deserves a comment explaining what's going on here (as mentioned before, I don't understand why this is necessary)

Reply:
See above. I'll add a comment